# 01 — Data Extraction & Inspection
**Project:** Brazilian E-Commerce Delivery & Satisfaction Analysis  
**Dataset:** Olist Brazilian E-Commerce (Kaggle)  
**Author:** Member 1 (ETL Lead)  
**Objective:** Load all raw CSV files, inspect structure, document schema, and confirm readiness for cleaning pipeline.

## 1. Imports & Configuration

In [1]:
import pandas as pd
import numpy as np
import os

pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)

RAW_PATH = '../data/raw/'

print('Libraries loaded successfully.')
print(f'Pandas version: {pd.__version__}')

Libraries loaded successfully.
Pandas version: 3.0.2


## 2. Load Raw CSV Files

In [2]:
# Load all 6 core Olist datasets
orders      = pd.read_csv(RAW_PATH + 'olist_orders_dataset.csv')
order_items = pd.read_csv(RAW_PATH + 'olist_order_items_dataset.csv')
customers   = pd.read_csv(RAW_PATH + 'olist_customers_dataset.csv')
reviews     = pd.read_csv(RAW_PATH + 'olist_order_reviews_dataset.csv')
sellers     = pd.read_csv(RAW_PATH + 'olist_sellers_dataset.csv')
products    = pd.read_csv(RAW_PATH + 'olist_products_dataset.csv')

datasets = {
    'orders':      orders,
    'order_items': order_items,
    'customers':   customers,
    'reviews':     reviews,
    'sellers':     sellers,
    'products':    products
}

print('All datasets loaded successfully.\n')
for name, df in datasets.items():
    print(f'{name:15s} → {df.shape[0]:,} rows × {df.shape[1]} columns')

All datasets loaded successfully.

orders          → 99,441 rows × 8 columns
order_items     → 112,650 rows × 7 columns
customers       → 99,441 rows × 5 columns
reviews         → 99,224 rows × 7 columns
sellers         → 3,095 rows × 4 columns
products        → 32,951 rows × 9 columns


## 3. Schema Inspection — Column Names & Data Types

In [3]:
for name, df in datasets.items():
    print(f'\n========== {name.upper()} ==========')
    print(df.dtypes.to_string())


========== ORDERS ==========
order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str

========== ORDER_ITEMS ==========
order_id                   str
order_item_id            int64
product_id                 str
seller_id                  str
shipping_limit_date        str
price                  float64
freight_value          float64

========== CUSTOMERS ==========
customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str

========== REVIEWS ==========
review_id                    str
order_id                     str
review_score               int64
review_comment_title         str
review_comment_message       str
review_creation_date   

## 4. Null Value Audit

In [4]:
print('NULL VALUE SUMMARY\n')
for name, df in datasets.items():
    null_counts = df.isnull().sum()
    null_pct    = (null_counts / len(df) * 100).round(2)
    summary     = pd.DataFrame({'null_count': null_counts, 'null_pct': null_pct})
    summary     = summary[summary['null_count'] > 0]
    if not summary.empty:
        print(f'--- {name} ---')
        print(summary.to_string())
        print()
    else:
        print(f'--- {name}: No nulls found ---\n')

NULL VALUE SUMMARY

--- orders ---
                               null_count  null_pct
order_approved_at                     160      0.16
order_delivered_carrier_date         1783      1.79
order_delivered_customer_date        2965      2.98

--- order_items: No nulls found ---

--- customers: No nulls found ---

--- reviews ---
                        null_count  null_pct
review_comment_title         87656     88.34
review_comment_message       58247     58.70

--- sellers: No nulls found ---

--- products ---
                            null_count  null_pct
product_category_name              610      1.85
product_name_lenght                610      1.85
product_description_lenght         610      1.85
product_photos_qty                 610      1.85
product_weight_g                     2      0.01
product_length_cm                    2      0.01
product_height_cm                    2      0.01
product_width_cm                     2      0.01



## 5. Sample Records

In [5]:
print('=== ORDERS (first 3 rows) ===')
display(orders.head(3))

print('\n=== ORDER ITEMS (first 3 rows) ===')
display(order_items.head(3))

print('\n=== REVIEWS (first 3 rows) ===')
display(reviews.head(3))

=== ORDERS (first 3 rows) ===


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00



=== ORDER ITEMS (first 3 rows) ===


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87



=== REVIEWS (first 3 rows) ===


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24


## 6. Basic Statistics — Numerical Columns

In [6]:
print('=== ORDERS describe() ===')
display(orders.describe())

print('\n=== ORDER ITEMS describe() ===')
display(order_items.describe())

print('\n=== REVIEWS describe() ===')
display(reviews.describe())

=== ORDERS describe() ===


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
count,99441,99441,99441,99441,99281,97658,96476,99441
unique,99441,99441,8,98875,90733,81018,95664,459
top,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2018-03-31 15:08:21,2018-02-27 04:31:10,2018-05-09 15:48:00,2018-05-14 20:02:44,2017-12-20 00:00:00
freq,1,1,96478,3,9,47,3,522



=== ORDER ITEMS describe() ===


,order_item_id,price,freight_value
count,112650.00,112650.00,112650.00
mean,1.20,120.65,19.99
std,0.71,183.63,15.81
min,1.00,0.85,0.00
25%,1.00,39.90,13.08
50%,1.00,74.99,16.26
75%,1.00,134.90,21.15
max,21.00,6735.00,409.68



=== REVIEWS describe() ===


,review_score
count,99224.00
mean,4.09
std,1.35
min,1.00
25%,4.00
50%,5.00
75%,5.00
max,5.00


## 7. Unique Value Counts — Key Categorical Columns

In [7]:
print('Order status distribution:')
print(orders['order_status'].value_counts().to_string())

print('\nReview score distribution:')
print(reviews['review_score'].value_counts().sort_index().to_string())

print('\nTop 10 product categories:')
print(products['product_category_name'].value_counts().head(10).to_string())

Order status distribution:
order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2

Review score distribution:
review_score
1    11424
2     3151
3     8179
4    19142
5    57328

Top 10 product categories:
product_category_name
cama_mesa_banho           3029
esporte_lazer             2867
moveis_decoracao          2657
beleza_saude              2444
utilidades_domesticas     2335
automotivo                1900
informatica_acessorios    1639
brinquedos                1411
relogios_presentes        1329
telefonia                 1134


## 8. Extraction Summary

| Dataset | Rows | Columns | Key Issues Identified |
|---|---|---|---|
| orders | 99,441 | 8 | Datetime columns as string, nulls in delivery dates |
| order_items | 112,650 | 7 | Multiple items per order (aggregation needed) |
| customers | 99,441 | 5 | Duplicate customer IDs possible |
| reviews | 99,224 | 7 | Nulls in review comment columns |
| sellers | 3,095 | 4 | Clean |
| products | 32,951 | 9 | Nulls in product category name |

